### ISSUE TITLE
System.OutOfMemoryException after playing for a while

### ISSUE THREAD

**Issue by gitskyflux** [vpinball/vpinball#304]
I hit a System.OutOfMemoryException often, typically during a long running game getting to or passing wizard modes, but sometimes in a shorter period of time like 5-10 minutes. I've done a lot of searching on the various forums and here, but I haven't seen anyone else mentioning this, so I can only assume this is specific to my environment. I'm running Windows 10 64-bit on an i7-9700k with 32 GB RAM and an RTX 4080. I have a widebody cabinet running two monitors, a PinDMDv3, 7.1 sound, and input/output controllers.

The gist of the play experience before the crash is consistent. First, the audio drops out for most or all mechanics (flippers, slingshots, etc.); however, sometimes a single component like a flipper or bumper will still produce audio and everything else drops out. Sometimes the audio will return briefly, but then drop out again before crashing. PinMAME audio continues without interruption or drop-out. Regardless, when the audio drops out, this is my cue the crash is coming soon, and I usually have between 10 and 30 seconds to bow out of the game if I want my score saved. Otherwise, VPX crashes.

Tables that reproduce this consistently:
Iron Man VPW Mod 1.0 (occurs in the shortest amount of time, probably under 10 minutes of gameplay, sometimes even faster)
Cirqus Voltaire Knorr 1.1 (usually during or after wizard mode)
Pirates of the Carribbean EBv1 (takes a while on this one, but consistent)

It has happened with other tables, and I imagine that if I play other tables long enough it will happen to them as well. I'm a developer, and this smells a lot like a classic memory leak. Unfortunately I don't develop anymore on this platform or I'd try to tackle fixing it myself.

I understand that 64-bit is not viable at this time, but I'm at the point of wanting to explore compiling the components that do not have x64 binaries ready, such as DOF and B2S. But before I go that route I wanted to see if anyone else has experienced this and just not reported it.

I'll do my best to dig up any other information as needed. I've been ignoring reporting this for at least 4 months, but that doesn't mean it started only 4 months ago.

**Crash dump**
The following 3 items reflect the most recent occurrence; however, most times P4 of the problem signature references libdmd and not System.Reactive.Core

Error: .NET Runtime
Application: VPinballX.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.OutOfMemoryException
   at System.Reactive.ExceptionHelpers..cctor()

Exception Info: System.TypeInitializationException
   at System.Reactive.ExceptionHelpers.Throw(System.Exception)
   at System.Reactive.Stubs.<.cctor>b__1(System.Exception)
   at System.Reactive.AnonymousObserver`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].OnErrorCore(System.Exception)
   at System.Reactive.ObserverBase`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].OnError(System.Exception)
   at System.Reactive.AutoDetachObserver`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].OnErrorCore(System.Exception)
   at System.Reactive.ObserverBase`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].OnError(System.Exception)
   at System.Reactive.Notification`1+OnErrorNotification[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].Accept(System.IObserver`1<System.__Canon>)
   at LibDmd.RenderGraph+<>c__DisplayClass62_1`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].<ObserveLatestOn>b__2(System.Action)
   at System.Reactive.Concurrency.Scheduler.<Schedule>b__45(System.Action`1<System.Action>, System.Action`1<System.Action`1<System.Action>>)
   at System.Reactive.Concurrency.Scheduler+<>c__DisplayClass50`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].<InvokeRec1>b__4d(System.__Canon)
   at System.Reactive.Concurrency.Scheduler+<>c__DisplayClass50`1+<>c__DisplayClass52[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].<InvokeRec1>b__4f(System.Reactive.Concurrency.IScheduler, System.__Canon)
   at System.Reactive.Concurrency.DefaultScheduler+<>c__DisplayClass1`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].<Schedule>b__0(System.Object)
   at System.Reactive.Concurrency.DefaultConcurrencyAbstractionLayer+<>c__DisplayClass1.<QueueUserWorkItem>b__0(System.Object)
   at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()

Error: Application Error
Faulting application name: VPinballX.exe, version: 10.7.3.0, time stamp: 0x63a5b2ce
Faulting module name: KERNELBASE.dll, version: 10.0.19041.2604, time stamp: 0xce23fdc3
Exception code: 0xe0434352
Fault offset: 0x0012e4b2
Faulting process id: 0x44c4
Faulting application start time: 0x01d94491964a3ce1
Faulting application path: D:\Games\Visual Pinball\VPinballX.exe
Faulting module path: C:\Windows\System32\KERNELBASE.dll
Report Id: f23d1ad2-aca5-474c-bf3c-8e20b79ffef1
Faulting package full name: 
Faulting package-relative application ID: 

Information: Windows Error Reporting
Fault bucket 1165715957163782832, type 5
Event Name: CLR20r3
Response: Not available
Cab Id: 0

Problem signature:
P1: VPinballX.exe
P2: 10.7.3.0
P3: 63a5b2ce
P4: System.Reactive.Core
P5: 2.2.5.0
P6: 53cee69b
P7: 1b2
P8: 0
P9: System.TypeInitialization
P10: 

**Versions**
 - VPX: 10.7.3 final 32-bit
 - DMD Extensions 1.10.2 as well as newer dmddevice.dll (crashes occur with and without update for PAC support)
 - Windows 10 21H2 64-bit
 - RTX 4080

**Comment by toxieainc** [comment]
Hmmm.. Maybe a DMDext leak? Cause 'at LibDmd.RenderGraph' appears in the dump

**Comment by gitskyflux** [comment]
I've since tried Iron Man without colorizing the DMD, and I ended up with the same issue.

**Comment by gitskyflux** [comment]
I tried running Iron Man with "Use external DMD" disabled. This gave me different results. The table audio died just as before, and then a few seconds later there was a pretty big series of frame drops over the span of maybe 250ms. VPX kept chugging along. I believe the area where frames were dropping was where it would typically crash. I was able to keep playing for another 10 minutes, but then the ROM audio started looping every 500ms. VPX kept going though, and it never crashed. PinMAME was still working correctly related to table mechanics, lighting, flashers, DOF, etc.

**Comment by vbousquet** [comment]
@gitskyflux DO you still have the problem with latest 10.8 ?
Also, some recent tables (XMen, Last Action Hero, Iron Man, Big Bang Bar,...) are targeted at the x64 version and may need it to run, or at least limit the texture size in your video settings.

**Comment by gitskyflux** [comment]
@vbousquet I tried the non-GL 10.8 beta x86, and though the problem persists, I was able to make it much farther and longer in Iron Man. I recently migrated to a dual x86+x64 configuration, and since running most tables in x64, including Iron Man, this has been an acceptable solution for my case. I'll leave it to you whether or not that means this issue can be closed. But in the end I doubt I will experience this issue again since x64 has been extremely reliable.
